tests/unit/test_gitlab_auth.py

```diff
@@ -83,3 +86,110 @@ def test_http_auth():
+
+
+@pytest.mark.parametrize(
+    "options,expected_private_token,expected_oauth_token,expected_job_token",
+    [
+        (
+            Namespace(
+                private_token="options-private-token",
+                oauth_token="options-oauth-token",
+                job_token="options-job-token",
+            ),
+            "options-private-token",
+            None,
+            None,
+        ),
+        (
+            Namespace(
+                private_token=None,
+                oauth_token="options-oauth-token",
+                job_token="option-job-token",
+            ),
+            None,
+            "options-oauth-token",
+            None,
+        ),
+        (
+            Namespace(
+                private_token=None, oauth_token=None, job_token="options-job-token"
+            ),
+            None,
+            None,
+            "options-job-token",
+        ),
+    ],
+)
+def test_get_auth_from_env_with_options(
+    options,
+    expected_private_token,
+    expected_oauth_token,
+    expected_job_token,
+):
+    cp = GitlabConfigParser()
+    cp.private_token = None
+    cp.oauth_token = None
+    cp.job_token = None
+
+    private_token, oauth_token, job_token = Gitlab._get_auth_from_env(options, cp)
+    assert private_token == expected_private_token
+    assert oauth_token == expected_oauth_token
+    assert job_token == expected_job_token
+
+
+@pytest.mark.parametrize(
+    "config,expected_private_token,expected_oauth_token,expected_job_token",
+    [
+        (
+            {
+                "private_token": "config-private-token",
+                "oauth_token": "config-oauth-token",
+                "job_token": "config-job-token",
+            },
+            "config-private-token",
+            None,
+            None,
+        ),
+        (
+            {
+                "private_token": None,
+                "oauth_token": "config-oauth-token",
+                "job_token": "config-job-token",
+            },
+            None,
+            "config-oauth-token",
+            None,
+        ),
+        (
+            {
+                "private_token": None,
+                "oauth_token": None,
+                "job_token": "config-job-token",
+            },
+            None,
+            None,
+            "config-job-token",
+        ),
+    ],
+)
+def test_get_auth_from_env_with_config(
+    config,
+    expected_private_token,
+    expected_oauth_token,
+    expected_job_token,
+):
+    options = Namespace(
+        private_token=None,
+        oauth_token=None,
+        job_token=None,
+    )
+    cp = GitlabConfigParser()
+    cp.private_token = config["private_token"]
+    cp.oauth_token = config["oauth_token"]
+    cp.job_token = config["job_token"]
+
+    private_token, oauth_token, job_token = Gitlab._get_auth_from_env(options, cp)
+    assert private_token == expected_private_token
+    assert oauth_token == expected_oauth_token
+    assert job_token == expected_job_token
```

Comment:
I'm asking a lot but would be nice to have a unit test which tests the ordering priority of the options.



1. Explicitly provided CLI arguments,
1. Environment variables,
1. Configuration files:
    -  explicitly defined config files:
    - via the --config-file CLI argument,
    - via the PYTHON_GITLAB_CFG environment variable,
    - user-specific config file,
    - system-level config file,
1. Environment variables always present in CI (CI_SERVER_URL, CI_JOB_TOKEN).

Reply:
Hehe yeah, I thought the same. But it got quite tedious and not easily doable in unit tests, so I've added functional tests where it's a more realistic scenario.

I think large parts of this are covered in different tests already though, just not the whole chain. Is there a big gap you noticed here that should be covered in this PR already?

Reply:
Yeah `functional` tests makes the most sense for me.

Not any big gaps that I can think of. I'll mark this as resolved.

Reply:
Maybe doing a test for at least the server `url` with the whole chain would make sense, one with `CI_SERVER_URL` present and one without.